File: unity-sdk/Packages/com.bugreporter.sdk/Runtime/BugReporter.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.SceneManagement;

namespace BugReporter
{
    public enum Severity { Low, Normal, High, Crash }

    /// <summary>
    /// Entry point. Call <see cref="Init"/> once (a RuntimeInitializeOnLoadMethod is a good place), then either
    /// let the floating button drive reports or call <see cref="Report"/> yourself.
    ///
    /// <code>
    /// BugReporter.Init(new BugReporterConfig {
    ///     ApiKey   = "br_live_xxx",
    ///     Endpoint = "https://us-central1-myproj.cloudfunctions.net/report",
    ///     Enabled  = Debug.isDebugBuild,
    /// });
    /// BugReporter.SetMetadata("innings", 2);
    /// </code>
    /// </summary>
    public static class BugReporter
    {
        private static BugReporterConfig _config;
        private static LogBuffer _logs;
        private static ReportSender _sender;
        private static readonly Dictionary<string, object> _metadata = new Dictionary<string, object>();
        private static readonly object _gameLock = new object();
        private static string _game = string.Empty;

        public static bool IsActive => _config != null && _config.Enabled;
        internal static BugReporterConfig Config => _config;
        internal static string CurrentGame { get { lock (_gameLock) return _game; } }

        public static void Init(BugReporterConfig config)
        {
            if (config == null) throw new ArgumentNullException(nameof(config));
            if (_config != null)
            {
                Debug.LogWarning("[BugReporter] Init called twice — ignoring the second call.");
                return;
            }
            if (!config.Enabled)
            {
                // Explicitly disabled (release build). Take no hooks, allocate nothing.
                _config = config;
                return;
            }
            // Pasted keys/URLs routinely arrive with stray whitespace or a newline — trim before validating,
            // or the server rejects every report with a confusing 401 (header present but malformed).
            config.ApiKey = config.ApiKey?.Trim();
            config.Endpoint = config.Endpoint?.Trim();
            if (string.IsNullOrEmpty(config.ApiKey) || string.IsNullOrEmpty(config.Endpoint))
            {
                Debug.LogError("[BugReporter] ApiKey and Endpoint are required — reporter stays off.");
                return;
            }
            if (!config.ApiKey.StartsWith("br_"))
            {
                Debug.LogError($"[BugReporter] ApiKey looks wrong ('{config.ApiKey.Substring(0, Math.Min(8, config.ApiKey.Length))}…') — it must start with br_. Get a fresh key from the dashboard (Settings → Rotate API key). Reporter stays off.");
                return;
            }

            if (string.IsNullOrEmpty(config.BuildVersion)) config.BuildVersion = Application.version;

            _config = config;
            lock (_gameLock) { _game = config.GameId?.Trim() ?? string.Empty; }
            _logs   = new LogBuffer(config.LogBufferSize, config.IncludeWarnings);
            _sender = ReportSender.Create();

            if (config.ShowReportButton) ReportOverlay.Create();

            Debug.Log($"[BugReporter] Ready — build {config.BuildVersion}, {config.LogBufferSize}-line log buffer.");
        }

        /// <summary>
        /// Attach live game state to whatever gets reported next. Overwrites the same key.
        /// Keep values small and JSON-friendly (string / number / bool).
        /// </summary>
        public static void SetMetadata(string key, object value)
        {
            if (!IsActive || string.IsNullOrEmpty(key)) return;
            lock (_metadata) { _metadata[key] = value; }
        }

        public static void ClearMetadata()
        {
            lock (_metadata) { _metadata.Clear(); }
        }

        /// <summary>
        /// Tag every subsequent report with the game the tester is currently in. Call this when a game scene
        /// loads (or the hub switches games); it sticks until the next call. One project/API key, many games —
        /// this is what lets the dashboard split them into separate lists instead of one mixed pile.
        /// </summary>
        public static void SetGame(string gameId)
        {
            lock (_gameLock) { _game = gameId?.Trim() ?? string.Empty; }
        }

        /// <summary>
        /// File a report. Captures a screenshot and the current log buffer, then uploads (with retry, and a
        /// disk queue if offline). Returns immediately — the work runs on a coroutine.
        /// </summary>
        public static void Report(string title, string description = null, Severity severity = Severity.Normal)
        {
            if (!IsActive || _sender == null) return;
            if (string.IsNullOrWhiteSpace(title)) title = "(no title)";

            Dictionary<string, object> meta;
            lock (_metadata) { meta = new Dictionary<string, object>(_metadata); }

            // The active scene's asset path (e.g. "Assets/_Games/CRICKET/Scene/Ground.unity") is preserved in
            // builds — the backend derives the game from the "_Games/<Game>/" segment, so no per-game wiring is
            // needed. GetActiveScene is main-thread only; Report runs on the main thread. `game` (from SetGame)
            // stays an optional override the server prefers when set.
            var active = SceneManager.GetActiveScene();
            string scenePath = string.IsNullOrEmpty(active.path) ? active.name : active.path;

            _sender.Submit(new ReportPayload
            {
                title        = title,
                description  = description ?? string.Empty,
                severity     = severity.ToString().ToLowerInvariant(),
                buildVersion = _config.BuildVersion,
                game         = CurrentGame,
                scene        = scenePath,
                logs         = _logs != null ? _logs.Dump() : string.Empty,
                metadata     = meta,
                device       = DeviceInfo.Capture(),
            });
        }

        /// <summary>Internal: the overlay asks for this when the user submits.</summary>
        internal static void ReportFromOverlay(string title, Severity severity) => Report(title, null, severity);
    }
}

File: flutter-sdk/bug_reporter/lib/src/bug_reporter.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:path_provider/path_provider.dart';

import 'capture.dart';
import 'clip_recorder.dart';
import 'config.dart';
import 'device.dart';
import 'log_buffer.dart';
import 'payload.dart';
import 'sender.dart';

enum Severity { low, normal, high, crash }

enum ReportOutcome {
  /// Delivered to the dashboard.
  sent,

  /// Couldn't reach the server; saved and sent on the next launch.
  queued,

  /// The server refused it (usually a wrong API key) — see [ReportResult.message].
  rejected,

  /// Not delivered and not saved.
  failed,

  /// The reporter is off (disabled, or init hasn't run / failed).
  inactive,
}

class ReportResult {
  const ReportResult(this.outcome, {this.message = ''});
  final ReportOutcome outcome;
  final String message;
}

/// Entry point: call [init] once before `runApp`, put [BugReporterOverlay] in `MaterialApp.builder`, and
/// testers get a floating 🐞 Report button. Reports carry a screenshot, the recent log lines, device and
/// build info, the current route, and whatever [setMetadata] / [setGame] / [setSession] recorded.
class BugReporter {
  BugReporter._();

  static BugReporterConfig? _config;
  static LogBuffer? _logs;
  static ReportSender? _sender;
  static ClipRecorder? _clip;
  static final DeviceInfoReader _device = DeviceInfoReader();
  static final Map<String, Object?> _metadata = {};
  static String _game = '';
  static String _session = '';
  static String _route = '';
  static String _buildVersion = '';

  /// True once [init] succeeded with `enabled: true`. The overlay listens to this.
  static final ValueNotifier<bool> active = ValueNotifier(false);

  static bool get isActive => active.value;
  static BugReporterConfig? get config => _config;

  /// Tracks the current route name so reports say which screen the tester was on.
  static final NavigatorObserver navigatorObserver = _RouteTracker();

  static Future<void> init(BugReporterConfig config) async {
    if (_config != null) {
      _say('init called twice — ignoring the second call.');
      return;
    }
    _config = config;
    if (!config.enabled) return;

    final apiKey = config.apiKey.trim(), endpoint = config.endpoint.trim();
    if (apiKey.isEmpty || endpoint.isEmpty) {
      _say('apiKey and endpoint are required — reporter stays off.');
      return;
    }
    if (!apiKey.startsWith('br_')) {
      _say("apiKey looks wrong ('${ReportSender.maskKey(apiKey)}') — it must start with br_. Reporter stays off.");
      return;
    }

    WidgetsFlutterBinding.ensureInitialized();
    _game = config.gameId?.trim() ?? '';
    _logs = LogBuffer(config.logBufferSize);
    _hookLogs(config);
    await _device.load();
    _buildVersion = config.buildVersion?.trim() ?? '';
    if (_buildVersion.isEmpty) {
      try {
        final info = await PackageInfo.fromPlatform();
        _buildVersion = info.buildNumber.isEmpty ? info.version : '${info.version}+${info.buildNumber}';
      } catch (_) {
        _buildVersion = 'unknown';
      }
    }

    _sender = ReportSender(
      endpoint: endpoint,
      apiKey: apiKey,
      queueDir: config.queueFailedReports && !kIsWeb ? _queueDir : null,
      log: _say,
    );
    unawaited(_sender!.flushQueue());

    if (config.recordClip) {
      _clip = ClipRecorder(
        seconds: config.clipSeconds, fps: config.clipFps, maxWidth: config.clipMaxWidth,
        quality: config.clipQuality, maxBytes: config.clipMaxBytes,
      )..start();
    }

    active.value = true;
    _say('Ready — build $_buildVersion, ${config.logBufferSize}-line log buffer, endpoint $endpoint, key ${ReportSender.maskKey(apiKey)}.');
  }

  static Future<Directory?> _queueDir() async {
    final base = await getApplicationSupportDirectory();
    return Directory('${base.path}/bugreporter-queue');
  }

  static void _hookLogs(BugReporterConfig config) {
    final previousFlutterError = FlutterError.onError;
    FlutterError.onError = (details) {
      _logs?.add('Error', '${details.exceptionAsString()}\n${details.stack ?? ''}'.trimRight());
      previousFlutterError?.call(details);
    };
    final dispatcher = WidgetsBinding.instance.platformDispatcher;
    final previousOnError = dispatcher.onError;
    dispatcher.onError = (error, stack) {
      _logs?.add('Exception', '$error\n$stack'.trimRight());
      return previousOnError?.call(error, stack) ?? false;
    };
    if (config.captureDebugPrint) {
      final previousPrint = debugPrint;
      debugPrint = (String? message, {int? wrapWidth}) {
        if (message != null) _logs?.add('Log', message);
        previousPrint(message, wrapWidth: wrapWidth);
      };
    }
  }

  /// Run the app inside a zone that also captures plain `print` calls and uncaught errors:
  /// `BugReporter.runGuarded(() => runApp(const MyApp()));`
  static void runGuarded(void Function() body) {
    runZonedGuarded(
      body,
      (error, stack) {
        _logs?.add('Exception', '$error\n$stack'.trimRight());
        FlutterError.presentError(FlutterErrorDetails(exception: error, stack: stack));
      },
      zoneSpecification: ZoneSpecification(print: (self, parent, zone, line) {
        _logs?.add('Log', line);
        parent.print(zone, line);
      }),
    );
  }

  /// Add a line to the log buffer yourself (e.g. from your own logger).
  static void log(String message, {String level = 'Log'}) => _logs?.add(level, message);

  /// Attach live app state to whatever gets reported next. Keep values small: string, number or bool.
  static void setMetadata(String key, Object? value) {
    if (key.isNotEmpty) _metadata[key] = value;
  }

  static void clearMetadata() => _metadata.clear();

  /// Tag every following report with the game/section the tester is in; the dashboard groups by it.
  static void setGame(String? gameId) => _game = gameId?.trim() ?? '';

  /// Tag every following report with the match/session id (the same id on every device in a multiplayer
  /// match) so reports from all devices show as one incident. Pass null when the match ends.
  static void setSession(String? sessionId) => _session = sessionId?.trim() ?? '';

  /// File a report: screenshot, logs, device info, then upload with retry (or save for later).
  static Future<ReportResult> report(String title, {String? description, Severity severity = Severity.normal}) async {
    final config = _config, sender = _sender;
    if (!isActive || config == null || sender == null) return const ReportResult(ReportOutcome.inactive);

    (Uint8List, Uint8List?)? shot;
    try {
      shot = await ScreenCapture.screenshot(quality: config.screenshotQuality, maxSide: config.screenshotMaxSide);
      if (shot == null) _say('No screenshot — is BugReporterOverlay in MaterialApp.builder? Sending without it.');
    } catch (e) {
      _say('Screenshot failed ($e) — sending the report without it.');
    }

    final clip = _clip?.packLatest();
    final payload = ReportPayload(
      title: title.trim().isEmpty ? '(no title)' : title.trim(),
      description: description?.trim() ?? '',
      severity: severity.name,
      buildVersion: _buildVersion,
      game: _game,
      session: _session,
      scene: _route,
      device: _device.current(),
      metadata: Map.of(_metadata),
      clipFps: clip != null ? _clip!.fps : 0,
    );

    final result = await sender.send(ReportParts(
      json: payload.toJson(),
      logs: _logs?.dump(),
      screenshot: shot?.$1,
      thumbnail: shot?.$2,
      clip: clip,
    ));
    return switch (result.status) {
      SendStatus.sent => ReportResult(ReportOutcome.sent, message: result.message),
      SendStatus.queued => ReportResult(ReportOutcome.queued, message: result.message),
      SendStatus.rejected => ReportResult(ReportOutcome.rejected, message: 'HTTP ${result.httpStatus}: ${result.message}'),
      SendStatus.failed => ReportResult(ReportOutcome.failed, message: result.message),
    };
  }

  static void _say(String message) {
    // Straight to the console: going through debugPrint would also copy these lines into the report's logs.
    debugPrintSynchronously('[BugReporter] $message');
  }

  @visibleForTesting
  static void debugReset() {
    _clip?.stop();
    _config = null;
    _logs = null;
    _sender = null;
    _clip = null;
    _metadata.clear();
    _game = _session = _route = _buildVersion = '';
    active.value = false;
  }

  @visibleForTesting
  static void debugActivate(BugReporterConfig config, ReportSender sender, {String buildVersion = 'test'}) {
    _config = config;
    _logs = LogBuffer(config.logBufferSize);
    _sender = sender;
    _buildVersion = buildVersion;
    active.value = true;
  }

  @visibleForTesting
  static String get debugRoute => _route;
}

class _RouteTracker extends NavigatorObserver {
  void _set(Route<dynamic>? route) {
    if (route is PageRoute) BugReporter._route = route.settings.name ?? route.runtimeType.toString();
  }

  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) => _set(route);
  @override
  void didReplace({Route<dynamic>? newRoute, Route<dynamic>? oldRoute}) => _set(newRoute);
  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) => _set(previousRoute);
  @override
  void didRemove(Route<dynamic> route, Route<dynamic>? previousRoute) => _set(previousRoute);
}
